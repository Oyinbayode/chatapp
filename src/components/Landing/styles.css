.chat-room-list {
  display: flex;
  flex-wrap: wrap;
  gap: 8px;
  width: 200px;
}

.chat-room-list li {
  height: 100px;
  background: var(--color-gray);
  flex: 1 1 calc(50% - 4px);
  border-radius: var(--border-radius);
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 16px;
  transition: font-size 0.2s ease-in-out;
}

.chat-room-list li:hover {
  background: var(--color-blue);
  color: var(--color-white);
  font-size: 18px;
}
