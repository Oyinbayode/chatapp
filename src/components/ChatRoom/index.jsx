import { Link, useParams } from "react-router-dom";
import "./styles.css";

const chatRoom = [
  {
    id: "general",
    title: "General Chat",
  },
];

function ChatRoom() {
  const params = useParams();

  const room = chatRoom.find((x) => x.id === params.id);
  return (
    <>
      <h2>{room.title}</h2>
      <div>
        <Link to="/">⬅️ Back to all rooms</Link>
      </div>
      <div className="messages-container">{/* TODO */}</div>
    </>
  );
}

export { ChatRoom };
