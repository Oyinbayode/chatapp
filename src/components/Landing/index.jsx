import { Link } from "react-router-dom";
import "./styles.css";

const chatRoom = [
  {
    id: "general",
    title: " ðŸ“£ General Chat",
  },
];

function Landing() {
  return (
    <>
      <h2>Choose a Chat Room</h2>
      <ul className="chat-room-list">
        {chatRoom.map((room) => (
          <li key={room.id}>
            <Link to={`/room/${room.id}`}>{room.title}</Link>
          </li>
        ))}
      </ul>
    </>
  );
}

export { Landing };
