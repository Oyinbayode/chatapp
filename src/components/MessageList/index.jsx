import React from "react";
import "./styles.css";

import { onSnapshot, collection, query } from "firebase/firestore";
import { db } from "../../services/firebase";
import { userContext } from "../../App";

function MessageList({ roomId }) {
  const containerRef = React.useRef(null);
  const { user } = React.useContext(userContext);

  const [messagesArray, setMessagesArray] = React.useState([]);

  React.useLayoutEffect(() => {
    if (containerRef.current) {
      containerRef.current.scrollTop = containerRef.current.scrollHeight;
    }
  });

  React.useEffect(() => {
    const unsubscribe = onSnapshot(
      query(collection(db, "chat-rooms", roomId, "messages")),
      (querySnapshot) => {
        querySnapshot.docChanges().forEach((change) => {
          if (change.type === "added") {
            setMessagesArray((prev) => [...prev, change.doc.data()]);
          }
        });
      }
    );
    return unsubscribe;
  }, [roomId]);

  return (
    <div className="message-list-container" ref={containerRef}>
      <ul className="message-list">
        {messagesArray.map((x) => (
          <Message
            key={x.id}
            message={x}
            isOwnMessage={x.displayName === user}
            time={x.createdAt}
          />
        ))}
      </ul>
    </div>
  );
}

function Message({ message, isOwnMessage }) {
  const { displayName, text } = message;
  return (
    <li className={["message", isOwnMessage && "own-message"].join(" ")}>
      <h4 className="sender">{isOwnMessage ? "You" : displayName}</h4>

      <div>{text}</div>
    </li>
  );
}

export { MessageList };
