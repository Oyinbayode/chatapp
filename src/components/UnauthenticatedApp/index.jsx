import { useState, useContext } from "react";
import { userContext } from "../../App";
import "./styles.css";

function UnauthenticatedApp() {
  const [userName, setUserName] = useState("");
  const { setUser } = useContext(userContext);

  const login = () => {
    if (userName === "") return;

    setUser(userName);
  };

  return (
    <>
      <h2>Enter a username to join a Chat Room!</h2>
      <div className="form">
        <input
          required
          className="form-input"
          type="text"
          placeholder="Enter UserName"
          value={userName}
          onChange={(e) => setUserName(e.target.value)}
        />
        <button
          disabled={userName === ""}
          onClick={() => login()}
          className="login"
        >
          Join
        </button>
      </div>
    </>
  );
}

export { UnauthenticatedApp };
