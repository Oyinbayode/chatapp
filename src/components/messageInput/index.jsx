import React from "react";
import { SendMessage } from "../../services/firebase";
import "./styles.css";
import { useDispatch } from "react-redux";
import { addUserMessages } from "../../redux/slices/userSlice";
import { userContext } from "../../App";

function MessageInput({ roomId }) {
  const { user } = React.useContext(userContext);

  const [value, setValue] = React.useState("");

  const dispatch = useDispatch();

  const handleChange = (event) => {
    setValue(event.target.value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    dispatch(addUserMessages({ userName: user, message: value }));
    SendMessage(roomId, user, value);
    setValue("");
  };

  return (
    <form onSubmit={handleSubmit} className="message-input-container">
      <input
        type="text"
        placeholder="Enter a message"
        value={value}
        onChange={handleChange}
        className="message-input"
        required
        minLength={1}
      />
      <button type="submit" disabled={value < 1} className="send-message">
        Send
      </button>
    </form>
  );
}

export { MessageInput };
