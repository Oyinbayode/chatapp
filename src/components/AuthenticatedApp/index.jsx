import { BrowserRouter, Route, Routes } from "react-router-dom";
import { ChatRoom } from "../ChatRoom";
import { Landing } from "../Landing";

function AuthenticatedApp() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Landing />} />
        <Route path="/room/:id" element={<ChatRoom />} />
      </Routes>
    </BrowserRouter>
  );
}

export { AuthenticatedApp };
