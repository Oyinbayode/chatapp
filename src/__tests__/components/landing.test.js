import React from "react";
import { render } from "@testing-library/react";
import { Landing } from "../../components/Landing";
import { MemoryRouter } from "react-router-dom";

describe("Landing", () => {
  it("should render the landing page", () => {
    const { container } = render(
      <MemoryRouter>
        <Landing />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
