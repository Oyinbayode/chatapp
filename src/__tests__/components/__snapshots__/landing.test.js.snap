// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing should render the landing page 1`] = `
<div>
  <h2>
    Choose a Chat Room
  </h2>
  <ul
    class="chat-room-list"
  >
    <li>
      <a
        href="/room/general"
      >
         ðŸ“£ General Chat
      </a>
    </li>
  </ul>
</div>
`;
